litho-core/src/main/java/com/facebook/litho/ComponentTree.java

```diff
@@ -1902,54 +1922,267 @@ protected LayoutState calculateLayoutState(
+          } catch (SecurityException e) {
+            /*
+              From {@link Process#THREAD_PRIORITY_DISPLAY}, some applications can not change
+              the thread priority to that of the main thread. This catches that potential error
+              and tries to set a lower priority.
+             */
+            targetThreadPriority += Process.THREAD_PRIORITY_LESS_FAVORABLE;
+          }
+        }
+      } else {
+        originalThreadPriority = THREAD_PRIORITY_INVALID;
+      }
+
+      final LayoutState layoutState;
+      try {
+        layoutState = super.get();
+      } catch (ExecutionException e) {
+        final Throwable cause = e.getCause();
+        if (cause instanceof RuntimeException) {
+          throw (RuntimeException) cause;
+        } else {
+          throw new RuntimeException(e.getMessage(), e);
+        }
+      } catch (InterruptedException | CancellationException e) {
+        throw new RuntimeException(e.getMessage(), e);
+      } finally {
+        if (originalThreadPriority != THREAD_PRIORITY_INVALID) {
+          // Reset the running thread's priority after we're unblocked.
+          Process.setThreadPriority(runningThreadId, originalThreadPriority);
+        }
+      }
+
+      // Creating a LayoutState gives it a refcount of 1, but subsequent get() calls to this Future
+      // must increment the refcount to ensure we don't release the LayoutState too early.
+      if (isFirstGet.getAndSet(false)) {
+        return layoutState;
+      }
+      try {
+        return layoutState.acquireRef();
+      } catch (IllegalStateException e) {
+        isReleased = true;
+        return null;
+      }
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+
+      LayoutStateFuture that = (LayoutStateFuture) o;
+
+      if (widthSpec != that.widthSpec) {
+        return false;
+      }
+      if (heightSpec != that.heightSpec) {
+        return false;
+      }
+      if (diffingEnabled != that.diffingEnabled) {
+        return false;
+      }
+      if (!context.equals(that.context)) {
+        return false;
+      }
+      if (root.getId() != that.root.getId()) {
+        // We only care that the root id is the same since the root is shallow copied before
+        // it's passed to us and will never be the same object.
+        return false;
+      }
+      if (diffingEnabled && previousLayoutState != null
+          ? !previousLayoutState.equals(that.previousLayoutState)
+          : that.previousLayoutState != null) {
+        // Only check against previousLayoutState if diffing is enabled.
+        return false;
+      }
+      if (treeProps != null ? !treeProps.equals(that.treeProps) : that.treeProps != null) {
+        return false;
+      }
+      return extraAttribution != null ? extraAttribution.equals(that.extraAttribution)
```

Comment:
Attribution shouldn't affect whether we re-use the layout: we may want to move where we use attribution to outside the future since we could have multiple reasons for waiting for the same layout. (attribution is just used for profiling and logging to understand why layouts are being created)

Reply:
It's used in the underlying call, so I won't move it out of the future for now, but I'll remove it from the .equals() check.